tutorial/practicals/nemo/4_nemo_openacc/README.md

```diff
@@ -0,0 +1,416 @@
+   within the outer, iteration loop of the mini-app. The script already
+   locates that loop:
+
+   ```python
+    # Find the outer, 'iteration' loop
+    tloop = None
+    for node in sched.children:
+        if isinstance(node, Loop) and node.loop_type == "tracers":
+            tloop = node
+            break
+    ```
+
+   Similar to what has been done in previous tutorials, you will need to
+   loop over the children of that loop, identify those that are loops of
+   the correct "levels" type, and transform them:
+   ```python
+    for child in tloop.loop_body.children:
+        if isinstance(node, Loop) and node.loop_type == "levels":
+	    ACC_KERNELS_TRANS.apply(node)
+   ```
+
+2. Use the supplied Makefile to run PSyclone and generate the transformed
+   code. If you examine the generated Fortran in `psy.f90` you should see
+   that ACC Kernels Directive nodes have been added to the Schedule,
+   e.g.:
+   ```fortran
+    DO jt = 1, it
+      !$ACC KERNELS
+      DO jk = 1, jpk
+        DO jj = 1, jpj
+          DO ji = 1, jpi
+            IF (tsn(ji, jj, jk) <= ztfreez(ji, jj) + 0.1D0) THEN
+              zice = 1.D0
+            ELSE
+              zice = 0.D0
+            END IF
+            zind(ji, jj, jk) = MAX(rnfmsk(ji, jj) * rnfmsk_z(jk), &
+	                       upsmsk(ji, jj), zice) * tmask(ji, jj, jk)
+            zind(ji, jj, jk) = 1 - zind(ji, jj, jk)
+          END DO
+        END DO
+      END DO
+      !$ACC END KERNELS
+      zwx(:, :, jpk) = 0.E0
+      zwy(:, :, jpk) = 0.E0
+      !$ACC KERNELS
+      DO jk = 1, jpk - 1
+        ...
+   ```
+
+Although this code will compile, run and produce the correct answers,
+its performance will be very poor. This is because we have only
+enclosed the loops over vertical levels within KERNELS regions and
+each of these regions acts as an implicit data region. This means that
+data will be copied from the host to the GPU before each region and
+then back again afterwards.
+
+If we are to get any sort of reasonabe performance then these memory
+copies must be eliminated by keeping data on the GPU for as long as
+possible. In order to do that we must first ensure that all computation
+is performed on the GPU.
+
+From the Fortran code fragment above, we can see that we have some
+array initialisation specified using Fortran array syntax. Since these
+are also loops and are writing to arrays, these too can be enclosed
+with KERNELS regions in order to perform them on the GPU.
+
+### 2. Enclose array assignments ###
+
+1. We will continue working with the `kernels_trans.py` script. It must
+   now be extended to identify those children of the iteration loop
+   that are array assignments. If this is the case they will be instances
+   of the `Assignment` class and will be marked as applying to an
+   array range:
+   ```python
+        # Enclose array assignments (implicit loops)
+        if isinstance(node, Assignment) and node.is_array_range:
```

Comment:
Use child again instead of node for the copy and pasters?

Reply:
You also don't declare Assignment in the script so probably should do.

Reply:
The apply for ACC_KERNELS_TRANS should be a list of nodes so add [child]

Reply:
I'm using `node` consistently now and have added the import of `Assignment`